src/main/java/com/azure/cosmos/kafka/connect/sink/BulkWriter.java

```diff
@@ -56,15 +54,38 @@ protected SinkWriteResponse writeCore(List<SinkRecord> sinkRecords) {
         }
 
         List<CosmosItemOperation> itemOperations = new ArrayList<>();
-        for (SinkRecord sinkRecord : sinkRecords) {
-            CosmosItemOperation cosmosItemOperation = CosmosBulkOperations.getUpsertItemOperation(
-                    sinkRecord.value(),
-                    this.getPartitionKeyValue(sinkRecord.value()),
-                    new SinkOperationContext(sinkRecord));
+        if (this.noDuplicates) {
+            HashMap<IdAndPartionKey, SinkRecord> uniqueItems = new HashMap<>();
+            for (SinkRecord sinkRecord : sinkRecords) {
+                IdAndPartionKey idAndPartionKey = new IdAndPartionKey(((Map<String, String>) sinkRecord.value()).get("id"), this.getPartitionKeyValue(sinkRecord.value()));
+                SinkRecord uniqueItem = uniqueItems.putIfAbsent(idAndPartionKey, sinkRecord);
+                if (uniqueItem != null && uniqueItem.timestamp() != null && sinkRecord.timestamp() != null
```

Comment:
what if there is no timestamp on sinkRecord? Do we not update the uniqueItem Map in that case?

Reply:
There should never be no timestamp because kafka guarantees a timestamp. If the user doesn't put one, kafka will. I put the null checks to adhere to the past unit tests which leave the timestamp as null.

Reply:
please use HaspMap.compute as replacement of putIfAbsent + conditional replace - that will reduce the necessary number of lookups in the map

Reply:
We were thinking of only doing the compression here and in a future pr adding the preserving order in the java sdk through a flag that is passed from here. That way the compression and preserving order logic stays separate and other connectors can benefit from the preserving order flag.

Reply:
Thanks changed it to a LinkedHashMap